forte/predictor.py

```diff
@@ -0,0 +1,185 @@
+                    feature = extractor.extract(pack, instance)
+                    feature_collection[tag] = feature
+
+            self.pack_pools.append(pack)
+            self.instance_pools.append(instance)
+            self.features_pools.append(feature_collection)
+
+            if len(self.features_pools) == self.batch_size:
+                yield self.convert(self.features_pools[:self.batch_size]), \
+                    self.pack_pools[:self.batch_size], \
+                    self.instance_pools[:self.batch_size]
+
+                self.features_pools.clear()
+                self.pack_pools.clear()
+                self.instance_pools.clear()
+
+        if not self.cross_pack and len(self.features_pools) > 0:
+            yield self.convert(self.features_pools), \
+                self.pack_pools, \
+                self.instance_pools
+            self.features_pools.clear()
+            self.pack_pools.clear()
+            self.instance_pools.clear()
+
+    def flush_batch(self) -> Dict:
+        if len(self.features_pools) > 0:
+            yield self.convert(self.features_pools), self.pack_pools, \
+                self.instance_pools
+            self.features_pools.clear()
+            self.pack_pools.clear()
+            self.instance_pools.clear()
+
+
+class Predictor(BaseProcessor):
+    '''This class will predict using the passed in model and add output
+    back to the datapack.
+    '''
+    def __init__(self, batch_size: int, model: Module,
+            predict_forward_fn: Callable, feature_resource: Dict,
+            cross_pack=True):
+        super().__init__()
+        self.feature_resource = feature_resource
+        self.batcher = Batcher(batch_size=batch_size,
+                                feature_resource=feature_resource,
+                                cross_pack=cross_pack)
+        self.model = model
+        self.predict_forward_fn = predict_forward_fn
+
+    def add_to_pack(self, predictions: Dict, packs: List[DataPack],
+                    instances: List[Annotation]):
+        for tag, preds in predictions.items():
+            for pred, pack, instance in zip(preds, packs, instances):
+                self.feature_resource['schemes'][tag]['extractor'].add_to_pack(
+                            pack, instance, pred)
+                pack.add_all_remaining_entries()
+
+    def predict(self, input_pack: DataPack):
+        for tensor_collection, packs, instances in \
+                                self.batcher.yield_batch(input_pack):
+            predictions = self.predict_forward_fn(self.model, tensor_collection)
+            self.add_to_pack(predictions, packs, instances)
+
+    def _process(self, input_pack: DataPack):
+        self.predict(input_pack)
+
+        # update the status of the jobs. The jobs which were removed from
+        # data_pack_pool will have status "PROCESSED" else they are "QUEUED"
+        q_index = self._process_manager.current_queue_index
```

Comment:
We have quite extensive tests here for BatchProcessor, maybe we should have similar tests?  (https://github.com/asyml/forte/blob/master/tests/forte/pipeline_test.py), or should we simply extend BatchProcessor?